packages/openapi-typescript/src/transform/schema-object.ts

```diff
@@ -142,10 +142,46 @@ export function transformSchemaObjectWithComposition(
   /** Collect oneOf/allOf/anyOf with Omit<> for discriminators */
   function collectCompositions(
     items: (SchemaObject | ReferenceObject)[],
+    required?: string[],
   ): ts.TypeNode[] {
     const output: ts.TypeNode[] = [];
     for (const item of items) {
-      const itemType = transformSchemaObject(item, options);
+      let itemType: ts.TypeNode;
+      // if this is a $ref, use WithRequired<X, Y> if parent specifies required properties
+      // (but only for valid keys)
+      if ("$ref" in item) {
+        itemType = transformSchemaObject(item, options);
+
+        const resolved = options.ctx.resolve<SchemaObject>(item.$ref);
+        if (
+          resolved &&
+          typeof resolved === "object" &&
+          "properties" in resolved
+        ) {
+          // don’t try and make keys required if the $ref doesn’t have them
+          const validRequired = (required ?? []).filter(
+            (key) => !!resolved.properties![key],
```

Comment:
This is probably brittle across multiple nested compositions, but that’s OK for now. I don’t think that’s a common scenario to need to specify `required` keys on a different schema object from a deeply-nested `allOf` chain